kyaml/fn/framework/framework.go

```diff
@@ -170,3 +321,120 @@ func (rl *ResourceList) Filter(api kio.Filter) error {
 }
+
+// SortItems sorts the ResourceList.items by apiVersion, kind, namespace and name.
+func (rl *ResourceList) SortItems() {
+	sort.Sort(sortRNodeObjects(rl.Items))
+}
+
+type sortRNodeObjects []*yaml.RNode
+
+func (o sortRNodeObjects) Len() int      { return len(o) }
+func (o sortRNodeObjects) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+func (o sortRNodeObjects) Less(i, j int) bool {
+	idi := o[i].GetResourceIdentifier()
+	idj := o[j].GetResourceIdentifier()
+	idStrI := fmt.Sprintf("%s %s %s %s", idi.GetAPIVersion(), idi.GetKind(), idi.GetNamespace(), idi.GetName())
+	idStrJ := fmt.Sprintf("%s %s %s %s", idj.GetAPIVersion(), idj.GetKind(), idj.GetNamespace(), idj.GetName())
+	return idStrI < idStrJ
+}
+
+// Run evaluates the function. input must be a resourceList in yaml format. An
+// updated resourceList will be returned.
+func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
+	rl, err := ParseResourceList(input)
+	if err != nil {
+		return nil, err
+	}
+	err = p.Process(rl)
+	if err != nil {
+		// If the error is not a Results type, we wrap the error as a Result.
+		if results, ok := err.(Results); ok {
+			rl.Results = append(rl.Results, results...)
+		} else if result, ok := err.(Result); ok {
+			rl.Results = append(rl.Results, &result)
+		} else if result, ok := err.(*Result); ok {
+			rl.Results = append(rl.Results, result)
+		} else {
+			rl.Results = append(rl.Results, ErrorResult(err))
+		}
+	}
+	yml, er := rl.ToYAML()
+	if er != nil {
+		return []byte(yml), er
+	}
+	if len(rl.Results) > 0 {
+		return []byte(yml), rl.Results
+	}
+	return []byte(yml), nil
+}
+
+type ErrMissingFnConfig struct{}
+
+func (ErrMissingFnConfig) Error() string {
+	return "unable to find the functionConfig in the resourceList"
+}
+
+// Chain chains a list of ResourceListProcessor as a single ResourceListProcessor.
+func Chain(processors ...ResourceListProcessor) ResourceListProcessor {
+	return ResourceListProcessorFunc(func(rl *ResourceList) error {
+		for _, processor := range processors {
+			if err := processor.Process(rl); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// ChainFunctions chains a list of ResourceListProcessorFunc as a single ResourceListProcessorFunc.
+func ChainFunctions(fns ...ResourceListProcessorFunc) ResourceListProcessorFunc {
+	return func(rl *ResourceList) error {
+		for _, fn := range fns {
+			if err := fn.Process(rl); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// GetPathAnnotation checks the path annotation first and then the legacy path
+// annotation. See: https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md#internalconfigkubernetesiopath
+func GetPathAnnotation(rn *yaml.RNode) string {
+	anno := rn.GetAnnotation(kioutil.PathAnnotation)
+	if anno == "" {
+		anno = rn.GetAnnotation(kioutil.LegacyPathAnnotation)
+	}
+	return anno
+}
+
+// GetIndexAnnotation checks the index annotation first and then the legacy index
```

Comment:
super nit: `GetIndexAnnotation gets the index from annotation "internal.config.kubernetes.io/index" or "config.kubernetes.io/index" (legacy).`  
 readability rationale: function docstring should explain what the function is for, not how the code is implemented.

Reply:
^ I agree. Why are function authors interested in these? We should explain here (including the "should not mutate" aspect) rather than sending to an external doc.